### DISCUSSION TITLE
How's the capability when we're trying to run into existing running device-farm service?

### DISCUSSION THREAD

**Discussion by ajsaprl** [AppiumTestDistribution/AppiumTestDistribution#1140] [Q&A] [answered]
I notice that with this capability we can run the automation with local appium server with device farm plugin:

> {
  "android": {
    "platformName": "android",
    "automationName": "UiAutomator2",
    "uiautomator2ServerInstallTimeout": 120000,
    "adbExecTimeout": 140000,
    "appActivity": "com.venture.test.MainActivity",
    "app": "./src/test/resources/apk/test/app-staging-debug-3.apk",
    "deviceName": "android",
    "noSign": true
  },
  "iOS": {
    "platformName": "iOS",
    "app": {
      "simulator": "./build/WordPress.zip"
    },
    "deviceName": "iPhone",
    "wdaLaunchTimeout": 80000,
    "wdaConnectionTimeout": 80000,
    "preventWDAAttachments": true,
    "usePrebuiltWDA": true,
    "clearSystemFiles": true,
    "shouldUseSingletonTestManager": false,
    "simpleIsVisibleCheck": true,
    "maxTypingFrequency": 10
  },
  "serverConfig" : {
    "server": {
      "use-plugins": ["device-farm"],
      "plugin": {
        "device-farm": {
          "platform": "android",
          "skipChromeDownload": true
        }
      }
    }
  }
}

But how is the capability when I already have appium device farm service running in my local/other host and want to run the script to the existing server?

**Comment by saikrishna321** [comment]
Take a look this please 

https://appium-device-farm-eight.vercel.app/remote-execution/

**Reply by ajsaprl** [reply]
Thank you for your response @saikrishna321 .
but i still can't understand how do i override appium server that already managed by ATDRunner to my own appium server?
also what's the Capabilities.json gonna be if i'm using remote execution? because when I remove:

>   "serverConfig" : {
    "server": {
      "use-plugins": ["device-farm"],
      "plugin": {
        "device-farm": {
          "platform": "android",
          "skipChromeDownload": true
        }
      }
    }
  }
  
  the test will be failed due to invalid capability format.

**Reply by saikrishna321** [reply]
You cannot connect to the running appium server. 

ATD manages its own server instance. 

What is your use-case why do u want to connect to ur server instance ?

**Reply by ajsaprl** [reply]
@saikrishna321 The use-case is simple: I want to run parallel test to my existing appium server. Can we do this only with spin up hub and node then hit appium test url execution to the hub (as mentioned here https://appium-device-farm-eight.vercel.app/remote-execution) without using ATD?

**Reply by saikrishna321** [reply]
ATD starts the server and manages it for parallel run. 

If u have the test and devices in the same machine you dont require remote-execution. 

If u have devices in multiple machines and test in another machine in this case you need remote-execution.

**Reply by saikrishna321** [reply]
Wihtout ATD also you do, start the server with device-farm and you need to send parallel request to the server. Device farm will take care if device management for each test.

**Reply by ajsaprl** [reply]
So I can achieve parallel execution to the remote server without using ATD?

**Reply by saikrishna321** [reply, marked as answer]
Yes! Please refer https://appium-device-farm-eight.vercel.app/remote-execution

**Reply by truyenaspire** [reply]
Hi @saikrishna321 can you share how to set parallel? I set update device in my local, don't need remote. 
I have 2 devices are connected and 10 tc in one Suite
```
{
  "server": {
    "port": 8888,
    "plugin": {
      "device-farm": {
        "platform": "android",
        "checkStaleDevicesIntervalMs": 5000,
        "maxSessions": 4
      }
    }
  }
}
```
but the test running in sequeue. Please give me some tips or I missed something in here?

**Reply by saikrishna321** [reply]
@truyenaspire what is the client ur using? Wdio or java-client ?

**Reply by truyenaspire** [reply]
@saikrishna321 thank for reply, I using java-client.
I just start use Appium Device Farm and try to setup parallel but it run only device for one time. 
Do we have any video or any article to share setup run parallel?

**Reply by saikrishna321** [reply]
@truyenaspire all you need do is trigger parallel request from client, if your using TestNG runner then you need to run in parallel thread. Thats exactly what AppiumTestDistribution project is doing. Can you check the project for ideas or use ATD.

**Reply by truyenaspire** [reply]
Hi @saikrishna321 For pom, I already added
```
  </dependencies>
    <dependency>
      <groupId>com.github.saikrishna321</groupId>
      <artifactId>AppiumTestDistribution</artifactId>
      <version>13.1.0</version>
    </dependency>
  </dependencies>
  <repositories>
    <repository>
      <id>jitpack.io</id>
      <url>https://jitpack.io</url>
    </repository>
  </repositories>
```

I have 2 device and I want to set 5 thread-count in TestNG like this:
![Screenshot 2023-12-22 at 15 04 43](https://github.com/AppiumTestDistribution/AppiumTestDistribution/assets/94594416/1f3cc203-329d-4fb4-a323-92b149b85890)
![Screenshot 2023-12-22 at 15 03 17](https://github.com/AppiumTestDistribution/AppiumTestDistribution/assets/94594416/be4d29ce-64f0-4585-b665-5243d5a8ebb1)
Is is enough or I need to something more
